src/runtime/objmodel.cpp

```diff
@@ -1705,8 +1711,37 @@ void setattrInternal(Box* obj, const std::string& attr, Box* val, SetattrRewrite
         // We don't need to to the invalidation stuff in this case.
         return;
     } else {
+        // Finally, check __setattr__
+        if (obj->cls->tp_setattr) {
+            rewrite_args = NULL;
+            REWRITE_ABORTED("");
+
+            int rtn = obj->cls->tp_setattr(obj, const_cast<char*>(attr.c_str()), val);
+            if (rtn)
+                throwCAPIException();
+            return;
+        }
+        Box* setattr = typeLookup(obj->cls, setattr_str, NULL);
+        if (setattr) {
+            rewrite_args = NULL;
+            REWRITE_ABORTED("");
+
+            // if we're dealing with a BoxedWrapperDescriptor wrapping
+            // PyObject_GenericSetAttr, skip calling __setattr__, as
+            // that will just re-enter us.
```

Comment:
Ah interesting... I think the issue is that PyObject_GenericSetAttr is not supposed to check for `__setattr__` (or `tp_setattr[o]`) at all.  This check covers the common case that PyObject_GenericSetAttr is called on an object where it is also the value of tp_setattro, but it's possible to call PyObject_GenericSetAttr on an object with a different tp_setattro, and then we'd have the same issue.  This could happen if an extension class inherits from an object and wants to add their own setattr that handles some but not all attribute names, and does something like:

```C++
int mysetattr(PyObject* obj, char* name, PyObject* value) {
  if (strcmp(name, "myattrname") == 0)
    return 0;
  return basecls->tp_setattr(obj, name, value);
}
```

I don't think this happens anywhere, but it did happen for for some other fields -- I ran into this with 79b2e9ccd835fb2cfeb47756a92295d4d17ef91b.  I hacked around it there, so I think the check you have here is good enough for now, but we'll have to fix these issues for real at some point :/